htlcswitch/link.go

```diff
@@ -2293,6 +2297,36 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 			}
 		}
 
+		// If we need to send out an Stfu, this would be the time to do
+		// so.
+		localOnRemote := l.channel.NumLocalUpdatesPendingOnRemote()
+		localOnLocal := l.channel.NumLocalUpdatesPendingOnLocal()
+		canSendStfu := localOnRemote == 0 && localOnLocal == 0
+		if l.quiescer.oweStfu() && canSendStfu {
+			handleErr := func(err error) {
+				l.log.Error(err.Error())
+				linkFailure := LinkFailureError{
+					code:             ErrInvalidUpdate,
+					FailureAction:    LinkFailureDisconnect,
+					PermanentFailure: false,
+					Warning:          true,
+				}
+				l.fail(linkFailure, err.Error())
+			}
+			oStfu, err := l.quiescer.sendStfu()
+			if err != nil {
+				handleErr(err)
+				return
+			}
+			oStfu.WhenSome(func(stfu lnwire.Stfu) {
+				err := l.cfg.Peer.SendMessage(false, &stfu)
```

Comment:
If we want to hide a bit more information about the `stfu` protocol from the link, then we can give the new state machine (with a layer of indirection) the ability to send messages itself, etc.

Reply:
I would generally like to avoid that. We are already encountering some race-y behavior due to a lack of linear message processing. I think this is another example of the type of logic we'd like to pack into a top level message handler interface that sits above the core subsystems that manage state but below the actual message router.

Reply:
Revisiting this, the other issue at hand is that the message has to be synchronized in with the rest of the link update handling loop. So we can't truly isolate the quiescence negotiation state and the link state.